### ISSUE TITLE
L2 regularization is performed twice!!

### ISSUE THREAD

**Issue by tommasocarraro** [rutgerswiselab/NCR#2]
Hi,
I checked the code many times and I think there is a conceptual error in it.

Basically, you perform L2 regularization two times at every optimizer step:
1. In your optimizer declaration (line 97 of file BaseRunner.py) you added the weight_decay parameter.
    `optimizer = torch.optim.Adam(model.parameters(), lr=self.learning_rate, weight_decay=self.l2_weight)`
2. Every time you compute the loss you add a manually computed L2 regularization term (line 177 of the same file). 
    `loss = output_dict['loss'] + model.l2() * self.l2_weight`

Is it possible this adds too much regularization?

Thank you!

**Comment by rutgerswiselab** [comment]
Dear @bmxitalia,

Thanks for the question. The l2 in line 177 is part of the loss function, while in line 97 it's part of the optimization algorithm provided by PyTorch's Adam optimizer. Of course, you can keep one of them if you want, that's also fine.

Besides, we see you have developed a Jupyter Notebook tutorial for the model, as you mentioned in our previous discussion in another thread. We really appreciate your efforts! We have added a pointer to your project in the README of this project. 

Thank you!

**Comment by tommasocarraro** [comment]
Hi @rutgerswiselab,
I think that when you put the weight decay on the Adam optimizer, PyTorch internally puts the L2 regularization term on the loss function. So, I think that in your case you performed L2 regularization twice.

Happy to hear that you enjoyed my demo on Neural Collaborative Filtering. Thank you for having posted my name on your repository! I feel grateful!

In the future, I will start a Ph.D. on the application of Neural-Symbolic integration to recommender systems. If there will be an occasion, I think there could be interesting collaborations.

Cheers,
Tommaso

**Comment by rutgerswiselab** [comment]
That's great. Look forward to possible collaborations.
